### PR TITLE
fix ConcurrentModificationException in KafkaIO

### PR DESCRIPTION
**Pull request by rangadi**
- after shutting down, wait for the threads to actually finish.
- handle a possible race between poll thread and close().

### CODE REVIEW

**Comment by rangadi** [comment]
R: @dhalperi & @tgroh . Please take a look when you get a chance.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 1]
```diff
@@ -1041,11 +1042,26 @@ public long getSplitBacklogBytes() {
...
+        // drain unread batch, this unblocks consumer thread. trying this in a loop to
+        // handle a small race where poll thread might try to enqueue after we drain.
+        consumer.wakeup();
+        availableRecordsQueue.poll();
+        try {
+          if (consumerPollThread.awaitTermination(10, TimeUnit.SECONDS)
+              && offsetFetcherThread.awaitTermination(10, TimeUnit.SECONDS)) {
+            break; // done
+          }
+        } catch (InterruptedException e) {
```
Comment (dhalperi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: should you set the interrupted bit here?

Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: I set the flag in two places where we are interrupted on thread not in our control.

**Comment by dhalperi** [comment]
- [ ] Can you confirm that you've run this with the `InProcessPipelineRunner` and the prior issues are not present?

LGTM once box is checked and tests pass. If you've done both and I haven't merged, ping me ;)

**Comment by rangadi** [comment]
R: @dhalperi , I could reproduce this by adding an extra cosumer.poll() after 'closed' flag was seen in consumerPollLoop(). I could not reproduce the problem without extra poll. 

With that, verified both that the fix works and in case the thread takes much longer, close() properly waits for it to shutdown with periodic warning log. 

Raghu.

**Comment by dhalperi** [comment]
LGTM, will merge soon.

**Pull request closed without merging**